bundles/azure-bundle/src/main/java/org/apache/gravitino/abs/credential/ADLSTokenProvider.java

```diff
@@ -0,0 +1,105 @@
+import com.azure.storage.file.datalake.DataLakeServiceClient;
+import com.azure.storage.file.datalake.DataLakeServiceClientBuilder;
+import com.azure.storage.file.datalake.models.UserDelegationKey;
+import com.azure.storage.file.datalake.sas.DataLakeServiceSasSignatureValues;
+import com.azure.storage.file.datalake.sas.PathSasPermission;
+import java.time.OffsetDateTime;
+import java.util.Map;
+import org.apache.gravitino.credential.ADLSTokenCredential;
+import org.apache.gravitino.credential.Credential;
+import org.apache.gravitino.credential.CredentialConstants;
+import org.apache.gravitino.credential.CredentialContext;
+import org.apache.gravitino.credential.CredentialProvider;
+import org.apache.gravitino.credential.PathBasedCredentialContext;
+import org.apache.gravitino.credential.config.ADLSCredentialConfig;
+
+/** Generates ADLS token to access ADLS data. */
+public class ADLSTokenProvider implements CredentialProvider {
+  private String storageAccountName;
+  private String tenantId;
+  private String clientId;
+  private String clientSecret;
+  private String endpoint;
+  private Integer tokenExpireSecs;
+
+  @Override
+  public void initialize(Map<String, String> properties) {
+    ADLSCredentialConfig adlsCredentialConfig = new ADLSCredentialConfig(properties);
+    this.storageAccountName = adlsCredentialConfig.storageAccountName();
+    this.tenantId = adlsCredentialConfig.tenantId();
+    this.clientId = adlsCredentialConfig.clientId();
+    this.clientSecret = adlsCredentialConfig.clientSecret();
+    this.endpoint = String.format("https://%s.dfs.core.windows.net", storageAccountName);
+    this.tokenExpireSecs = adlsCredentialConfig.tokenExpireInSecs();
+  }
+
+  @Override
+  public void close() {}
+
+  @Override
+  public String credentialType() {
+    return CredentialConstants.ADLS_TOKEN_CREDENTIAL_PROVIDER_TYPE;
+  }
+
+  @Override
+  public Credential getCredential(CredentialContext context) {
+    if (!(context instanceof PathBasedCredentialContext)) {
+      return null;
+    }
+
+    TokenCredential tokenCredential =
+        new ClientSecretCredentialBuilder()
+            .tenantId(tenantId)
+            .clientId(clientId)
+            .clientSecret(clientSecret)
+            .build();
+
+    DataLakeServiceClient dataLakeServiceClient =
+        new DataLakeServiceClientBuilder()
+            .endpoint(endpoint)
+            .credential(tokenCredential)
+            .buildClient();
+
+    OffsetDateTime start = OffsetDateTime.now();
+    OffsetDateTime expiry = start.plusSeconds(tokenExpireSecs);
+    UserDelegationKey userDelegationKey = dataLakeServiceClient.getUserDelegationKey(start, expiry);
+
+    // Define Path SAS permissions, read, write, delete, list
+    PathSasPermission permission = PathSasPermission.parse("rwdl");
+    DataLakeServiceSasSignatureValues signatureValues =
+        new DataLakeServiceSasSignatureValues(expiry, permission);
```

Comment:
Could Azure supports generates a token contains the privilage information to the specific location like S3Token and GCSToken?

Reply:
`PathSasPermission` is used to specify permissions.

If we want to specify a particular location, we can do so when creating the SAS token.

```java
String sasToken =
        new DataLakeSasImplUtil(sasSignatureValues, container, path, true)
            .generateUserDelegationSas(key, accountName, Context.NONE);
```

However, it only accepts a single path.  
Please refer to [`DataLakeSasImplUtil`](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/implementation/util/DataLakeSasImplUtil.java#L34).

In this case, the path should align with the warehouse path.

Could I do something like `properties.get(IcebergConstants.warehouse)` in `ADLSTokenProvider.initialize` to retrieve the warehouse path?

Reply:
1. It's critical to restrict the access path in azure tokens for security.
2. We could get the path information in `PathBasedCredentialContext`, as it only supports one location, we could add a check for the path, throw exception if there are multi paths, since in most cases, there should be only one path.

Reply:
Updated